lib/AssemblyUtil/FastaToAssembly.py

```diff
@@ -100,15 +104,40 @@ def _import_fasta_mass(self, params):
                 json.dump(ao, f)
 
         # save to WS and return
-        assembly_infos = self._save_assembly_objects(
-            params[_WSID],
-            [p[_ASSEMBLY_NAME] for p in params[_INPUTS]],
-            assobjects
-        )
+        assembly_infos = []
+        assembly_names = [p[_ASSEMBLY_NAME] for p in params[_INPUTS]]
+        # generator to process files in bach. Avoid memory issues
+        for assobject_batch, assembly_name_batch in self._assembly_objects_generator(
+            assobjects, assembly_names
+        ):
+            assembly_infos.extend(
+                self._save_assembly_objects(params[_WSID], assembly_name_batch, assobject_batch)
+            )
+
         for out, ai in zip(output, assembly_infos):
             out['upa'] = _upa(ai)
         return output
 
+    def _run_parallel_import_fasta_mass(self, params):
+        threads = max(int(os.cpu_count() * min(_SYSTEM_UTILIZATION, 1)), 1)
```

Comment:
Ah, what is catalog service and where can I do environmental parameter setting?

Reply:
https://github.com/kbase/catalog
serves most of https://ci.kbase.us/#catalog/

To set the params you need to be a catalog admin. I can probably set some if you need to experiment.

See the recent discussion in `#kbase_coders` for how they work

Reply:
Let's say we set default threads/number of workers as 5. We probably don't need maximum thread count at all?

Reply:
I'd like to be able to adjust the threads / cpu and maximum thread count dynamically as our hardware changes. 5 looks best for running in a 2 cpu docker container on perlmutter login (right?) but may not be for other hardware (like the job queue on KBase hardward)

Reply:
I see. 

Ah, 5 look best for running `kb-sdk` in a 2 cpu docker container on my local. I believe on perlmutter login will be faster.

Reply:
Could you please help me set the following 2 catalog parameters and I can finish the rest of logic/experiment:)

- default workers = 5
- default maximum workers=10

Reply:
CI catalog?

Reply:
Actually I'm confused by those args. If we're just setting an absolute worker count, what's the point of also setting a maximum?

Reply:
nah, this line of code will be replaced by a helper function called _get_workers(), which will deal with workers and maximum workers logic inside.

Reply:
I'm not understanding the answer? If a catalog admin can just set workers=5, when would maximum_workers ever come into play?

Reply:
We had the same thought: https://github.com/kbaseapps/AssemblyUtil/pull/89#discussion_r1332247609

Let's just set default workers. If hardware changes, developers can adjust workers to see what fits the best.

Reply:
The issue I have there is that this may run on hardware with different numbers of processors. The KBase job nodes are not necessarily homogeneous. That's why I'm suggesting 2 variables - a threads / CPU variable, AU can scale up on beefier hardware, and a max threads variable, so it doesn't overload external systems and can be set to play nice with other jobs on the same node

Reply:
I have seen `--numWorkers` and `--maxNumWorkers` params used in most of cloud services. Not sure if they set it for the same purpose, but what you are suggesting makes sense to me. How can we set those? I assume I need to somehow parse them in the script?

Reply:
how are they used in those cloud services?

See the discussion starting here for how catalog vars work, but they should end up in the `config` dict passed into the Impl file constructor IIUC: https://kbase.slack.com/archives/C4E7KUGTD/p1694558822102569

Reply:
I set in entrypoint.sh like what Shane suggested: https://kbase.slack.com/archives/C4E7KUGTD/p1694561082878559
Please let me know if you think differently.

Reply:
What shane is talking about is how the env vars get transmitted from the catalog service to the app. They shouldn't actually be hard coded into the entrypoint script. You shouldn't have to make any changes other than get the variables from the `config` in the impl file constructor

Reply:
Where can I set them before getting the variables from the config?

Reply:
They have to be set in the catalog by a catalog admin. I can do that for you for CI, but maybe ask in #sysadmin if you can also have admin privs so you can test it yourself. Not sure how strict they are about that

Reply:
If we pass in `KB_ADMIN_AUTH_TOKEN` to the callback server, secure params can be retrieved. 
https://github.com/kbase/JobRunner/blob/callback_module/JobRunner/CatalogCache.py#L30-L34

My understanding is based on the `callback_module` branch in JobRunner. Other branches have different logic.